datafusion/expr/src/expr_rewriter/mod.rs

```diff
@@ -336,12 +338,14 @@ impl NamePreserver {
 }
 
 impl SavedName {
-    /// Ensures the name of the rewritten expression is preserved
+    /// Ensures the qualified name of the rewritten expression is preserved
     pub fn restore(self, expr: Expr) -> Result<Expr> {
         let Self(original_name) = self;
         match original_name {
-            Some(name) => expr.alias_if_changed(name),
```

Comment:
I seems like we could potentially also remove `Expr::alias_if_changed` which doesn't properly account for qualifiers 🤔 

The only other use of it seems to be in https://github.com/apache/datafusion/blob/e60318553d6ac36b2d07466acebef861fde2936e/datafusion/expr/src/expr_rewriter/mod.rs#L287 which itself is only used in tests 🤔

Reply:
`alias_if_changed` is being used by [substrait](https://github.com/apache/datafusion/blob/9d819e1e237f5b10a197bc0a6548c1ad9c4c467c/datafusion/substrait/src/logical_plan/consumer.rs#L315). Maybe we can review this later.

 `rewrite_preserving_name` has been removed.